R/raincloudPlots.R

```diff
@@ -0,0 +1,908 @@
+    plotInProgress  <- plotInProgress + addCaption + captionPosition
+  }
+
+  # Assign to inputPlot
+  return(plotInProgress)
+}  # End .rainFillPlot()
+
+
+
+# .rainSetPalettes() ----
+.rainSetPalettes <- function(dataset, options) {
+
+  fillTitle <- if (options$secondaryFactor == "") options$primaryFactor else options$secondaryFactor
+  paletteFill <- if (options$secondaryFactor != "" || options$colorAnyway) {
+    jaspGraphs::scale_JASPfill_discrete(options$colorPalette, name = fillTitle)
+  } else {
+    NULL
+  }
+
+  paletteColor <- if (options$covariate != "") {
+    if (is.factor(dataset$covariate)) {
+      jaspGraphs::scale_JASPcolor_discrete(  options$covariatePalette, name = options$covariate)
+    } else {
+      jaspGraphs::scale_JASPcolor_continuous(options$covariatePalette, name = options$covariate)
+    }
+  } else {
+    if (options$secondaryFactor != "" || options$colorAnyway) {
+      jaspGraphs::scale_JASPcolor_discrete(  options$colorPalette,     name = options$secondaryFactor)
+    } else {
+      NULL
+    }
+  }
+
+  return(list(fill = paletteFill, color = paletteColor))
+}  # End .rainSetPalettes()
+
+
+
+# .rainInfoFactorCombinations() ----
+# Calculates info to determine colors & geom orientation
+.rainInfoFactorCombinations <- function(inputDataset, inputPlot, extractColor = TRUE) {
+  onlyFactors    <- inputDataset[c("primaryFactor", "secondaryFactor")]
+  # Extract used Fill colors from plot
+  # https://stackoverflow.com/questions/11774262/how-to-extract-the-fill-colours-from-a-ggplot-object
+
+  if (extractColor) {  # So that we can re-use it in .rainFillTable()
+
+    onlyFactors$color <- tryCatch(
+      {
+        ggplot2::ggplot_build(inputPlot)$data[[1]]["fill"]$fill  # Requires secondaryFactor or colorAnyway
+      },
+      error = function(e) {                                      # Thus error handling, but not used further
+        return("black")
+      },
+      warning = function(w) {
+        return("black")
+      }
+    )
+
+  } else {
+    onlyFactors$color <- "black"
+  }
+
+  # In the following possibleCombis <- expand.grid() secondaryFactor first because
+  # then structure will match order in which ggplot accesses the clouds:
+  # for each level of primaryFactor, we get the levels of secondaryFactor
+  # (as opposed to: for each level of secondaryFactor, the levels of primaryFactor)
+  possibleCombis <- expand.grid(
+    secondaryFactor = levels(onlyFactors$secondaryFactor), primaryFactor = levels(onlyFactors$primaryFactor)
+  )
+
+  possibleCombis$rowId <- 1:nrow(possibleCombis)  # Id is necessary as merge will scramble order of possibleCombis
+  observedCombis <- merge(possibleCombis, onlyFactors)
+  observedCombis <- observedCombis[order(observedCombis$rowId), ] # Re-order rows according to id
+  observedCombis <- observedCombis[ , c("primaryFactor", "secondaryFactor", "color", "rowId")]  # Re-order columns
+
+  uniqueCombis   <- dplyr::count(observedCombis, pick(everything()))
```

Comment:
Is this the only use case for `dplyr`? If yes, I would drop it as a dependency and do this is in a different way.

Reply:
After your comment, I had dropped this line and thus the dplyr dependency.
Instead I crafted multiple lines of base R as a work-around.
However, we later noticed that plotting would crash when there were factor levels with spaces in their names: e.g. factor hairColor with two levels; "dark brunette" and "light blonde".

Then, I tracked the problem down to my base R work-around.
I tried to fix it with just base R, but then decided that using this simple dplyr line was the best solution.
Because it is tidy and did not cost me much more time to implement.

I asked @boutinb if that is fine and he said its okay.